pkg/controller/persistentvolume/persistentvolume_framework_test.go

```diff
@@ -0,0 +1,462 @@
+//   - testSyncClaimError - calls syncClaim on the first claim in initialClaims
+//                          and expects an error to be returned.
+//   - testSyncVolume - calls syncVolume on the first volume in initialVolumes.
+//   - any custom function for specialized tests.
+// The test then contains list of volumes/claims that are expected at the end
+// of the test.
+type controllerTest struct {
+	// Name of the test, for logging
+	name string
+	// Initial content of controller volume cache.
+	initialVolumes []*api.PersistentVolume
+	// Expected content of controller volume cache at the end of the test.
+	expectedVolumes []*api.PersistentVolume
+	// Initial content of controller claim cache.
+	initialClaims []*api.PersistentVolumeClaim
+	// Expected content of controller claim cache at the end of the test.
+	expectedClaims []*api.PersistentVolumeClaim
+	// Function to call as the test.
+	test testCall
+}
+
+type testCall func(ctrl *PersistentVolumeController, reactor *volumeReactor, test controllerTest) error
+
+const testNamespace = "default"
+const versionConflictError = "VersionError"
+
+var noVolumes []*api.PersistentVolume
+var noClaims []*api.PersistentVolumeClaim
+
+// volumeReactor is a core.Reactor that simulates etcd and API server. It
+// stores:
+// - Latest version of claims volumes saved by the controller.
+// - Queue of all saves (to simulate "volume/claim updated" events). This queue
+//   contains all intermediate state of an object - e.g. a claim.VolumeName
+//   is updated first and claim.Phase second. This queue will then contain both
+//   updates as separate entries.
+// - Number of changes since the last call to volumeReactor.syncAll().
+// - Optionally, volume and claim event sources. When set, all changed
+//   volumes/claims are sent as Modify event to these sources. These sources can
+//   be linked back to the controller watcher as "volume/claim updated" events.
+type volumeReactor struct {
+	volumes              map[string]*api.PersistentVolume
+	claims               map[string]*api.PersistentVolumeClaim
+	changedObjects       []interface{}
+	changedSinceLastSync int
+	ctrl                 *PersistentVolumeController
+	volumeSource         *framework.FakeControllerSource
+	claimSource          *framework.FakeControllerSource
+	lock                 sync.Mutex
+}
+
+// React is a callback called by fake kubeClient from the controller.
+// In other words, every claim/volume change performed by the controller ends
+// here.
+// This callback checks versions of the updated objects and refuse those that
+// are too old (simulating real etcd).
+// All updated objects are stored locally to keep track of object versions and
+// to evaluate test results.
+// All updated objects are also inserted into changedObjects queue and
+// optionally sent back to the controller via its watchers.
+func (r *volumeReactor) React(action core.Action) (handled bool, ret runtime.Object, err error) {
```

Comment:
A lot of this feels really generically useful - maybe we can just leave a TODO here that with a little polish it could become more useful for other unit tests (key word: "unit") to test against a fake etcd more completely?

@lavalamp FYI